File: vitaltrack-backend/app/api/v1/items.py
"""
VitalTrack Backend - Item Routes
CRUD operations for inventory items
"""

from typing import Optional

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import and_, or_, select

from app.api.deps import DB, CurrentUser, Pagination
from app.models import ActivityActionType, ActivityLog, Category, Item
from app.schemas import (
    ItemCreate,
    ItemList,
    ItemResponse,
    ItemStats,
    ItemUpdate,
    StockUpdate,
    SuccessResponse,
)


router = APIRouter(prefix="/items", tags=["Items"])


# =============================================================================
# LIST ITEMS
# =============================================================================
@router.get(
    "",
    response_model=ItemList,
    summary="List all items",
)
async def list_items(
    db: DB,
    current_user: CurrentUser,
    category_id: Optional[str] = Query(None, alias="categoryId"),
    is_active: Optional[bool] = Query(None, alias="isActive"),
    is_critical: Optional[bool] = Query(None, alias="isCritical"),
    low_stock_only: Optional[bool] = Query(False, alias="lowStockOnly"),
    out_of_stock_only: Optional[bool] = Query(False, alias="outOfStockOnly"),
    search: Optional[str] = Query(None, max_length=100),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100, alias="pageSize"),
) -> ItemList:
    """
    Get all items for the current user with optional filters.
    
    - **categoryId**: Filter by category
    - **isActive**: Filter by active status
    - **isCritical**: Filter by critical flag
    - **lowStockOnly**: Only show items below minimum stock
    - **outOfStockOnly**: Only show items with zero quantity
    - **search**: Search in name, description, and brand
    """
    query = select(Item).where(Item.user_id == current_user.id)
    
    # Apply filters
    if category_id:
        query = query.where(Item.category_id == category_id)
    
    if is_active is not None:
        query = query.where(Item.is_active == is_active)
    
    if is_critical is not None:
        query = query.where(Item.is_critical == is_critical)
    
    if low_stock_only:
        query = query.where(
            and_(
                Item.quantity > 0,
                Item.quantity < Item.minimum_stock,
            )
        )
    
    if out_of_stock_only:
        query = query.where(Item.quantity <= 0)
    
    if search:
        search_term = f"%{search}%"
        query = query.where(
            or_(
                Item.name.ilike(search_term),
                Item.description.ilike(search_term),
                Item.brand.ilike(search_term),
            )
        )
    
    # Order by name
    query = query.order_by(Item.name)
    
    # Get total count
    count_result = await db.execute(
        select(Item.id).where(Item.user_id == current_user.id)
    )
    total = len(count_result.all())
    
    # Apply pagination
    offset = (page - 1) * page_size
    query = query.offset(offset).limit(page_size)
    
    result = await db.execute(query)
    items = result.scalars().all()
    
    return ItemList(
        items=[ItemResponse.model_validate(item) for item in items],
        total=total,
    )


@router.get(
    "/stats",
    response_model=ItemStats,
    summary="Get item statistics",
)
async def get_item_stats(
    db: DB,
    current_user: CurrentUser,
) -> ItemStats:
    """
    Get inventory statistics for dashboard.
    Returns counts matching frontend DashboardStats interface.
    """
    # Get items
    result = await db.execute(
        select(Item).where(
            Item.user_id == current_user.id,
            Item.is_active == True,
        )
    )
    items = result.scalars().all()
    
    total_items = len(items)
    out_of_stock = sum(1 for item in items if item.quantity <= 0)
    low_stock = sum(1 for item in items if 0 < item.quantity < item.minimum_stock)
    critical_items = sum(1 for item in items if item.is_critical)
    
    # Get total categories count
    from sqlalchemy import func
    cat_result = await db.execute(
        select(func.count()).select_from(Category).where(
            Category.user_id == current_user.id
        )
    )
    total_categories = cat_result.scalar() or 0
    
    # Get pending orders count (pending or received status)
    from app.models import Order, OrderStatus
    pending_result = await db.execute(
        select(func.count()).select_from(Order).where(
            Order.user_id == current_user.id,
            Order.status.in_([OrderStatus.PENDING, OrderStatus.RECEIVED])
        )
    )
    pending_orders_count = pending_result.scalar() or 0
    
    return ItemStats(
        total_items=total_items,
        total_categories=total_categories,
        out_of_stock=out_of_stock,
        low_stock=low_stock,
        critical_items=critical_items,
        pending_orders_count=pending_orders_count,
    )


@router.get(
    "/needs-attention",
    response_model=ItemList,
    summary="Get items needing attention",
)
async def get_items_needing_attention(
    db: DB,
    current_user: CurrentUser,
) -> ItemList:
    """
    Get items that are out of stock or low on stock.
    
    Results are ordered with out-of-stock items first,
    then critical items, then by quantity.
    """
    result = await db.execute(
        select(Item)
        .where(
            Item.user_id == current_user.id,
            Item.is_active == True,
            or_(
                Item.quantity <= 0,
                Item.quantity < Item.minimum_stock,
            ),
        )
        .order_by(
            Item.quantity,  # Out of stock first
            Item.is_critical.desc(),  # Critical items next
            Item.name,
        )
    )
    items = result.scalars().all()
    
    return ItemList(
        items=[ItemResponse.model_validate(item) for item in items],
        total=len(items),
    )


# =============================================================================
# GET SINGLE ITEM
# =============================================================================
@router.get(
    "/{item_id}",
    response_model=ItemResponse,
    summary="Get item by ID",
)
async def get_item(
    item_id: str,
    db: DB,
    current_user: CurrentUser,
) -> ItemResponse:
    """
    Get a single item by ID.
    """
    result = await db.execute(
        select(Item).where(
            Item.id == item_id,
            Item.user_id == current_user.id,
        )
    )
    item = result.scalar_one_or_none()
    
    if not item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Item not found",
        )
    
    return ItemResponse.model_validate(item)


# =============================================================================
# CREATE ITEM
# =============================================================================
@router.post(
    "",
    response_model=ItemResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new item",
)
async def create_item(
    data: ItemCreate,
    db: DB,
    current_user: CurrentUser,
) -> ItemResponse:
    """
    Create a new inventory item.
    
    - **categoryId**: Category to assign item to (required)
    - **name**: Item name (required)
    - **quantity**: Current stock level
    - **minimumStock**: Reorder threshold
    - **isCritical**: Mark as critical equipment
    """
    # Verify category exists and belongs to user
    cat_result = await db.execute(
        select(Category).where(
            Category.id == data.category_id,
            Category.user_id == current_user.id,
        )
    )
    if not cat_result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Category not found",
        )
    
    item = Item(
        user_id=current_user.id,
        category_id=data.category_id,
        name=data.name,
        description=data.description,
        quantity=data.quantity,
        unit=data.unit,
        minimum_stock=data.minimum_stock,
        expiry_date=data.expiry_date,
        brand=data.brand,
        notes=data.notes,
        supplier_name=data.supplier_name,
        supplier_contact=data.supplier_contact,
        purchase_link=data.purchase_link,
        image_uri=data.image_uri,
        is_active=data.is_active,
        is_critical=data.is_critical,
        local_id=data.local_id,
    )
    db.add(item)
    await db.flush()
    
    # Log activity
    activity = ActivityLog(
        user_id=current_user.id,
        action=ActivityActionType.ITEM_CREATE,
        item_name=item.name,
        item_id=item.id,
        details=f"Created with {item.quantity} {item.unit}",
    )
    db.add(activity)
    
    await db.commit()
    await db.refresh(item)
    
    return ItemResponse.model_validate(item)


# =============================================================================
# UPDATE ITEM
# =============================================================================
@router.put(
    "/{item_id}",
    response_model=ItemResponse,
    summary="Update an item",
)
async def update_item(
    item_id: str,
    data: ItemUpdate,
    db: DB,
    current_user: CurrentUser,
) -> ItemResponse:
    """
    Update an existing item.
    
    All fields are optional - only provided fields will be updated.
    """
    result = await db.execute(
        select(Item).where(
            Item.id == item_id,
            Item.user_id == current_user.id,
        )
    )
    item = result.scalar_one_or_none()
    
    if not item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Item not found",
        )
    
    # Track changes for activity log
    changes = []
    
    # Verify category if changing
    if data.category_id and data.category_id != item.category_id:
        cat_result = await db.execute(
            select(Category).where(
                Category.id == data.category_id,
                Category.user_id == current_user.id,
            )
        )
        if not cat_result.scalar_one_or_none():
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Category not found",
            )
        item.category_id = data.category_id
        changes.append("category changed")
    
    # Update fields
    if data.name is not None:
        item.name = data.name
    if data.description is not None:
        item.description = data.description
    if data.quantity is not None and data.quantity != item.quantity:
        changes.append(f"qty: {item.quantity} → {data.quantity}")
        item.quantity = data.quantity
    if data.unit is not None:
        item.unit = data.unit
    if data.minimum_stock is not None:
        item.minimum_stock = data.minimum_stock
    if data.expiry_date is not None:
        item.expiry_date = data.expiry_date
    if data.brand is not None:
        item.brand = data.brand
    if data.notes is not None:
        item.notes = data.notes
    if data.supplier_name is not None:
        item.supplier_name = data.supplier_name
    if data.supplier_contact is not None:
        item.supplier_contact = data.supplier_contact
    if data.purchase_link is not None:
        item.purchase_link = data.purchase_link
    if data.image_uri is not None:
        item.image_uri = data.image_uri
    if data.is_active is not None:
        item.is_active = data.is_active
    if data.is_critical is not None:
        item.is_critical = data.is_critical
    
    # Log activity
    activity = ActivityLog(
        user_id=current_user.id,
        action=ActivityActionType.ITEM_UPDATE,
        item_name=item.name,
        item_id=item.id,
        details=", ".join(changes) if changes else "Updated",
    )
    db.add(activity)
    
    await db.commit()
    await db.refresh(item)
    
    return ItemResponse.model_validate(item)


# =============================================================================
# UPDATE STOCK ONLY
# =============================================================================
@router.patch(
    "/{item_id}/stock",
    response_model=ItemResponse,
    summary="Update item stock",
)
async def update_stock(
    item_id: str,
    data: StockUpdate,
    db: DB,
    current_user: CurrentUser,
) -> ItemResponse:
    """
    Quick update of item stock level.
    
    Use this for simple stock adjustments.
    """
    result = await db.execute(
        select(Item).where(
            Item.id == item_id,
            Item.user_id == current_user.id,
        )
    )
    item = result.scalar_one_or_none()
    
    if not item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Item not found",
        )
    
    old_quantity = item.quantity
    item.quantity = data.quantity
    
    # Log activity
    activity = ActivityLog(
        user_id=current_user.id,
        action=ActivityActionType.STOCK_UPDATE,
        item_name=item.name,
        item_id=item.id,
        details=f"Stock: {old_quantity} → {data.quantity} {item.unit}",
    )
    db.add(activity)
    
    await db.commit()
    await db.refresh(item)
    
    return ItemResponse.model_validate(item)


# =============================================================================
# DELETE ITEM
# =============================================================================
@router.delete(
    "/{item_id}",
    response_model=SuccessResponse,
    summary="Delete an item",
)
async def delete_item(
    item_id: str,
    db: DB,
    current_user: CurrentUser,
) -> SuccessResponse:
    """
    Delete an inventory item.
    """
    result = await db.execute(
        select(Item).where(
            Item.id == item_id,
            Item.user_id == current_user.id,
        )
    )
    item = result.scalar_one_or_none()
    
    if not item:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Item not found",
        )
    
    item_name = item.name
    
    # Log activity before deletion
    activity = ActivityLog(
        user_id=current_user.id,
        action=ActivityActionType.ITEM_DELETE,
        item_name=item_name,
        details="Item deleted",
    )
    db.add(activity)
    
    await db.delete(item)
    await db.commit()
    
    return SuccessResponse(message=f"Item '{item_name}' deleted")

File: vitaltrack-mobile/services/items.ts
/**
 * VitalTrack Mobile - Items API Service
 * CRUD operations for items with filtering and stats
 */

import { api } from './api';
import type { Item, DashboardStats } from '@/types';

// Response types
interface ItemsListResponse {
  items: Item[];
  total: number;
}

interface NeedsAttentionResponse {
  items: Item[];
  total: number;
  outOfStockCount: number;
  lowStockCount: number;
}

// Query parameters for filtering items
interface ItemsQueryParams {
  categoryId?: string;
  search?: string;
  isActive?: boolean;
  isCritical?: boolean;
  lowStock?: boolean; // Legacy alias for lowStockOnly
  outOfStock?: boolean; // Legacy alias for outOfStockOnly
  lowStockOnly?: boolean;
  outOfStockOnly?: boolean;
  page?: number;
  pageSize?: number;
  skip?: number; // Legacy offset, converted to page when limit/pageSize is present
  limit?: number; // Legacy alias for pageSize; values over 100 fetch all pages
}

// Create item request
interface CreateItemRequest {
  categoryId: string;
  name: string;
  description?: string;
  quantity?: number;
  unit?: string;
  minimumStock?: number;
  expiryDate?: string;
  brand?: string;
  notes?: string;
  supplierName?: string;
  supplierContact?: string;
  purchaseLink?: string;
  imageUri?: string;
  isCritical?: boolean;
}

// Update item request
interface UpdateItemRequest {
  categoryId?: string;
  name?: string;
  description?: string;
  quantity?: number;
  unit?: string;
  minimumStock?: number;
  expiryDate?: string;
  brand?: string;
  notes?: string;
  supplierName?: string;
  supplierContact?: string;
  purchaseLink?: string;
  imageUri?: string;
  isCritical?: boolean;
  isActive?: boolean;
  version: number;
}

// Stock update request
interface StockUpdateRequest {
  quantity: number;
  version: number;
}

// Drop undefined and empty-string fields. Keeps falsy values like 0 and false.
// Avoids sending empty `expiryDate: ""` (fails date parsing) or `purchaseLink: ""`
// (fails URL validator) while still allowing required fields and zero quantities.
function stripEmpty<T extends Record<string, unknown>>(data: T): Partial<T> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(data)) {
    if (value === undefined) continue;
    if (typeof value === 'string' && value.trim() === '') continue;
    out[key] = value;
  }
  return out as Partial<T>;
}

// Build query string from params
const MAX_PAGE_SIZE = 100;

function buildQueryString(params?: ItemsQueryParams): string {
  if (!params) return '';
  
  const queryParts: string[] = [];
  const requestedPageSize = params.pageSize ?? params.limit;
  const pageSize =
    requestedPageSize !== undefined
      ? Math.max(1, Math.min(MAX_PAGE_SIZE, requestedPageSize))
      : undefined;
  const page =
    params.page ??
    (params.skip !== undefined && pageSize !== undefined
      ? Math.floor(params.skip / pageSize) + 1
      : undefined);
  const lowStockOnly = params.lowStockOnly ?? params.lowStock;
  const outOfStockOnly = params.outOfStockOnly ?? params.outOfStock;
  
  if (params.categoryId) queryParts.push(`categoryId=${encodeURIComponent(params.categoryId)}`);
  if (params.search) queryParts.push(`search=${encodeURIComponent(params.search)}`);
  if (params.isActive !== undefined) queryParts.push(`isActive=${params.isActive}`);
  if (params.isCritical !== undefined) queryParts.push(`isCritical=${params.isCritical}`);
  if (lowStockOnly !== undefined) queryParts.push(`lowStockOnly=${lowStockOnly}`);
  if (outOfStockOnly !== undefined) queryParts.push(`outOfStockOnly=${outOfStockOnly}`);
  if (page !== undefined) queryParts.push(`page=${page}`);
  if (pageSize !== undefined) queryParts.push(`pageSize=${pageSize}`);
  
  return queryParts.length > 0 ? `?${queryParts.join('&')}` : '';
}

export const itemService = {
  /**
   * Get all items with optional filters
   */
  async getAll(params?: ItemsQueryParams): Promise<ItemsListResponse> {
    if (
      params?.limit !== undefined &&
      params.limit > MAX_PAGE_SIZE &&
      params.page === undefined &&
      params.skip === undefined
    ) {
      const items: Item[] = [];
      let total = 0;
      let page = 1;

      while (true) {
        const queryString = buildQueryString({
          ...params,
          limit: undefined,
          skip: undefined,
          page,
          pageSize: MAX_PAGE_SIZE,
        });
        const response = await api.get<ItemsListResponse>(`/items${queryString}`);
        items.push(...response.items);
        total = response.total;

        if (items.length >= total || response.items.length === 0) break;
        page += 1;
      }

      return { items, total };
    }

    const queryString = buildQueryString(params);
    return api.get<ItemsListResponse>(`/items${queryString}`);
  },

  /**
   * Get dashboard statistics
   */
  async getStats(): Promise<DashboardStats> {
    return api.get<DashboardStats>('/items/stats');
  },

  /**
   * Get items that need attention (low stock or out of stock)
   */
  async getNeedsAttention(): Promise<NeedsAttentionResponse> {
    return api.get<NeedsAttentionResponse>('/items/needs-attention');
  },

  /**
   * Get single item by ID
   */
  async getById(id: string): Promise<Item> {
    return api.get<Item>(`/items/${id}`);
  },

  /**
   * Create new item
   * Strip undefined and empty-string optional fields so the backend validators
   * (URL, date) don't reject them.
   */
  async create(data: CreateItemRequest): Promise<Item> {
    const cleaned = stripEmpty(data as unknown as Record<string, unknown>);
    return api.post<Item>('/items', cleaned);
  },

  /**
   * Update existing item
   */
  async update(id: string, data: UpdateItemRequest): Promise<Item> {
    const cleaned = stripEmpty(data as unknown as Record<string, unknown>);
    return api.put<Item>(`/items/${id}`, cleaned);
  },

  /**
   * Quick stock update (requires version for OCC)
   */
  async updateStock(id: string, quantity: number, version: number): Promise<Item> {
    return api.patch<Item>(`/items/${id}/stock`, { quantity, version });
  },

  /**
   * Delete item
   */
  async delete(id: string): Promise<{ message: string }> {
    return api.delete<{ message: string }>(`/items/${id}`);
  },
};

export type { 
  ItemsQueryParams, 
  CreateItemRequest, 
  UpdateItemRequest, 
  StockUpdateRequest,
  NeedsAttentionResponse 
};
